pymc_marketing/clv/models/shifted_beta_geo.py

```diff
@@ -0,0 +1,192 @@
+import types
+from typing import Optional, Union
+
+import numpy as np
+import pandas as pd
+import pymc as pm
+from pymc import str_for_dist
+from pytensor.tensor import TensorVariable
+from xarray import DataArray, Dataset
+
+from pymc_marketing.clv.models import CLVModel
+
+
+class BaseShiftedBetaGeoModel(CLVModel):
+    def _process_priors(self, alpha_prior, beta_prior):
+        if alpha_prior is None:
+            alpha_prior = pm.HalfFlat.dist()
+        else:
+            assert alpha_prior.eval().shape == ()
+        if beta_prior is None:
+            beta_prior = pm.HalfFlat.dist()
+        else:
+            assert beta_prior.eval().shape == ()
+
+        assert alpha_prior is not beta_prior
+
+        alpha_prior.str_repr = types.MethodType(str_for_dist, alpha_prior)
+        beta_prior.str_repr = types.MethodType(str_for_dist, beta_prior)
+
+        return alpha_prior, beta_prior
+
+
+class ShiftedBetaGeoModel(BaseShiftedBetaGeoModel):
+    pass  # Implement Model that expects data aggregated by cohorts
+
+
+class ShiftedBetaGeoModelIndividual(BaseShiftedBetaGeoModel):
+    """Shifted Beta Geometric model
+
+    Model for customer behavior in a contractual setting. It assumes that:
+      * At the end of each period, a customer has a probability `theta` of renewing the contract
+        and `1-theta` of cancelling
+      * The probability `theta` does not change over time for a given customer
+      * The probability `theta` varies across customers according to a beta prior distribution
+        with hyperparameters `alpha` and `beta`.
+
+    based on [1]_.
+
+    Parameters
+    ----------
+    customer_id: array_like
+        Customer labels. The same value should be used for each observation
+        coming from the same customer.
+    t_churn: array_like
+        Time at witch customer cancelled contract (starting at 0). It should equal
+        t_end for users that have not cancelled by the end of the observation period
+    t_end: array_like
```

Comment:
No, I was struggling with names. Are you suggesting `T` instead of `t_end` or `t_churn`? If you have a suggestion for both even better.

Reply:
`t_churn` is fine, but maybe `t_periods` or `t_observed` instead of `t_end`? "End" is somewhat misleading because these customers may still be active. `T` works as well since it's equivalent across model settings; just comes down to if you want to use this parameter name to highlight the fact this is discrete time model.

Reply:
Yeah I'll stick with `T`. Thanks for the input